src/kv/encryptor.h

```diff
@@ -2,144 +2,85 @@
-      // that is lower than the given version (e.g. if encryption_keys contains
-      // two keys for version 0 and 10 then the key associated with version 0
-      // is used for version [0..9] and version 10 for versions 10+)
-      auto search = std::upper_bound(
-        encryption_keys.rbegin(),
-        encryption_keys.rend(),
-        version,
-        [](kv::Version a, EncryptionKey const& b) { return b.version <= a; });
-
-      if (search == encryption_keys.rend())
-      {
-        throw std::logic_error(fmt::format(
-          "TxEncryptor: encrypt version is not valid: {}", version));
-      }
-
-      return search->key;
+      // IV is function of seqno, term and snapshot so that:
+      // - Same seqno across rollbacks does not reuse IV
+      // - Snapshots do not reuse IV
+      // - If all nodes execute the _same_ tx (or generate the same snapshot),
+      // the same IV will be used
+
+      // Note that only the first 31 bits of the term are used for the IV which
+      // is acceptable for a live CCF as 2^31 elections will take decades, even
+      // with an election timeout as low as 1 sec.
+
+      hdr.set_iv_seq(tx_id.version);
+      hdr.set_iv_term(tx_id.term);
+      hdr.set_iv_snapshot(is_snapshot);
     }
 
-  protected:
-    // Encryption keys are set when TxEncryptor object is created and are used
-    // to determine which key to use for encryption/decryption when
-    // committing/deserialising depending on the version
+  public:
+    TxEncryptor(std::shared_ptr<T> secrets, bool is_recovery_ = false) :
+      ledger_secrets(secrets),
+      is_recovery(is_recovery_)
+    {}
 
-    struct EncryptionKey : KeyInfo
+    void disable_recovery() override
     {
-      crypto::KeyAesGcm key;
-    };
-
-    std::list<EncryptionKey> encryption_keys;
-    size_t iv_id = 0;
-
-    virtual void record_compacted_keys(const std::list<KeyInfo>&){};
+      is_recovery = false;
+    }
 
-  public:
-    TxEncryptor(const std::list<KeyInfo>& existing_keys)
+    size_t get_header_length() override
     {
-      // Create map of existing encryption keys from the recorded ledger secrets
-      for (auto const& s : existing_keys)
-      {
-        encryption_keys.emplace_back(TxEncryptor::EncryptionKey{
-          {s.version, s.raw_key}, crypto::KeyAesGcm(s.raw_key)});
-      }
+      return S::RAW_DATA_SIZE;
     }
 
-    /**
-     * Encrypt data and return serialised GCM header and cipher.
-     *
-     * @param[in]   plain             Plaintext to encrypt
-     * @param[in]   additional_data   Additional data to tag
-     * @param[out]  serialised_header Serialised header (iv + tag)
-     * @param[out]  cipher            Encrypted ciphertext
-     * @param[in]   version           Version used to retrieve the corresponding
-     * encryption key
-     * @param[in]   is_snapshot       Indicates that the entry is a snapshot (to
-     * avoid IV re-use)
-     */
```

Comment:
Why have we lost these docs?

Reply:
Because I re-wrote this class from scratch and forgot to add this back - well spotted! I've added it back.